lib/elixir/lib/kernel.ex

```diff
@@ -1971,10 +1971,17 @@ defmodule Kernel do
   In case any entry returns `nil`, its key will be removed
   and the deletion will be considered a success.
   """
-  @spec pop_in(Access.t, nonempty_list(term)) :: {term, Access.t}
+  @spec pop_in(Access.container, nonempty_list(term)) :: {term, Access.container}
   def pop_in(data, keys)
-  def pop_in(nil, [h | _]), do: Access.pop(nil, h)
-  def pop_in(data, keys), do: do_pop_in(data, keys)
+
+  def pop_in(nil, [head | _]) do
+    raise ArgumentError,
```

Comment:
the thing is that Dialyzer is complaining because `Access.pop(nil, h)` does not type succeeds.
I guess the best we can do is to create a public function in Kernel that raises. WDYT